components/util/model.jsx

```diff
@@ -0,0 +1,60 @@
+import React from 'react';
+import { immutableMerge } from './index.js';
+
+export function Model({ initialState, methods }) {
+  const listeners = [];
+  const notifyListeners = nextState => listeners.forEach(fn => fn(nextState));
+
+  const model = {
+    state: Object.freeze(initialState),
+    subscribe: fn => listeners.push(fn),
+    unsubscribe: fn => listeners.splice(listeners.indexOf(fn), 1),
+  };
+
+  Object.keys(methods).forEach((method) => {
+    model[method] = (...args) => {
+      const updates = methods[method].apply(model, args);
+      if (updates && typeof updates === 'object') {
+        model.state = immutableMerge(model.state, updates);
+        notifyListeners(model.state);
+      }
+      // could also check here to see if fn returns a promise and then wait
+      // until it resolves to merge and notifyListeners. that would make
+      // async methods really nice to work with.
+      // ...should this return  something? maybe return `updates`?
```

Comment:
This is a case where it would not place nicely with redux, since you cannot subscribe to specific methods being called.

So while this may be fine with redux:
```js
counterModel.subscribe(function(nextState) {
  store.dispatch('COUNT_CHANGED', nextState)
});
```

This would not:

```js
storeModel.subscribe(function(nextState) {
  store.dispatch('STORE_CHANGED', nextState)
});
```

...because the listener runs when either the search or products change, and you likely care about `search` and `loadProducts` separately rather than merged together.